File: sdk/runanywhere-swift/Sources/RunAnywhere/Public/RunAnywhere.swift
//
//  RunAnywhere.swift
//  RunAnywhere SDK
//
//  The main entry point for the RunAnywhere SDK.
//  Two-phase initialization is owned by commons (rac_sdk_init.h):
//    * Phase 1 → rac_sdk_init_phase1_proto (validate + state init)
//    * Phase 2 → rac_sdk_init_phase2_proto (auth/refresh, device
//      registration, model assignments, telemetry flush, model discovery)
//    * HTTP retry → rac_sdk_retry_http_proto
//  Swift retains only the parts that cannot move into C++:
//    * Task.detached spawning + Swift-side initialization state
//    * Keychain SDK params persistence (Apple-specific)
//    * MainActor platform-plugin/callback registration
//    * URLSession HTTP transport implementation and adapter configuration
//

import Darwin
import Foundation
import os

/// The RunAnywhere SDK - Single entry point for on-device AI
public enum RunAnywhere {

    // MARK: - Internal State Management

    private struct SDKState: @unchecked Sendable {
        var initParams: SDKInitParams?
        var currentEnvironment: SDKEnvironment?
        var isInitialized = false
        var hasCompletedServicesInit = false
        var hasCompletedHTTPSetup = false
        var httpSetupApplicable = true
        var servicesInitTask: Task<Void, Error>?
    }

    private static let state = OSAllocatedUnfairLock<SDKState>(initialState: SDKState())

    /// Serializes Phase 1 core initialization so concurrent `initialize()`
    /// callers cannot both enter the C++ bridge setup path.
    private static let coreInitQueue = DispatchQueue(label: "com.runanywhere.sdk.coreInit")

    /// Internal init params storage.
    internal static var initParams: SDKInitParams? { state.withLock { $0.initParams } }
    internal static var currentEnvironment: SDKEnvironment? { state.withLock { $0.currentEnvironment } }
    internal static var isInitializedFlag: Bool { state.withLock { $0.isInitialized } }

    /// Track if services initialization is complete (makes API calls O(1) after first use).
    internal static var hasCompletedServicesInit: Bool { state.withLock { $0.hasCompletedServicesInit } }
    /// Track if HTTP/auth setup succeeded (separate from core services so auth can be retried on reconnect).
    internal static var hasCompletedHTTPSetup: Bool { state.withLock { $0.hasCompletedHTTPSetup } }
    internal static var isLocalOnlyMode: Bool {
        guard let rawValue = getenv("RUNANYWHERE_SWIFT_LOCAL_ONLY") else {
            return false
        }
        let value = String(cString: rawValue).lowercased()
        return value == "1" || value == "true" || value == "yes"
    }

    // MARK: - SDK State

    /// Check if SDK is initialized (Phase 1 complete).
    public static var isInitialized: Bool { isInitializedFlag }

    /// Check if services are fully ready (Phase 2 complete)
    public static var areServicesReady: Bool { hasCompletedServicesInit }

    /// Check if SDK is active and ready for use
    public static var isActive: Bool {
        state.withLock { $0.isInitialized && $0.initParams != nil }
    }

    /// Current SDK version
    public static var version: String { SDKConstants.version }

    /// Current environment (nil if not initialized)
    public static var environment: SDKEnvironment? { currentEnvironment }

    /// Device ID (Keychain-persisted, survives reinstalls)
    /// Resolved by commons via the device-identity chain
    /// (secure_get → vendor ID → freshly synthesized UUID).
    public static var deviceId: String { CppBridge.Device.persistentId }

    // MARK: - Event Access

    /// Access to all SDK events for subscription-based patterns
    public static var events: EventBus { EventBus.shared }

    // MARK: - Authentication Info (Production/Staging only)

    /// Get current user ID from authentication
    public static func getUserId() -> String? { CppBridge.State.userId }

    /// Get current organization ID from authentication
    public static func getOrganizationId() -> String? { CppBridge.State.organizationId }

    /// Check if currently authenticated
    public static var isAuthenticated: Bool { CppBridge.Auth.isAuthenticated }

    /// Check if device is registered with backend
    public static func isDeviceRegistered() -> Bool { CppBridge.Device.isRegistered }

    // MARK: - SDK Reset (Testing)

    /// Reset SDK state (for testing purposes)
    public static func reset() async {
        let logger = SDKLogger(category: "RunAnywhere.Reset")
        logger.info("Resetting SDK state...")

        let taskToCancel = state.withLock { lockedState -> Task<Void, Error>? in
            let task = lockedState.servicesInitTask
            lockedState = SDKState()
            return task
        }
        taskToCancel?.cancel()

        await CppBridge.shutdown()
        CppBridge.State.shutdown()

        logger.info("SDK state reset completed")
    }

    // MARK: - SDK Initialization

    /// Initialize the RunAnywhere SDK.
    /// Phase 1 runs synchronously; Phase 2 spawns in a detached Task.
    public static func initialize(
        apiKey: String? = nil,
        baseURL: String? = nil,
        environment: SDKEnvironment = .development
    ) throws {
        let params: SDKInitParams
        if environment == .development {
            params = SDKInitParams(forDevelopmentWithAPIKey: apiKey ?? "")
        } else {
            params = try SDKInitParams(
                apiKey: apiKey ?? "",
                baseURL: baseURL ?? "",
                environment: environment
            )
        }
        try performCoreInit(with: params, startBackgroundServices: true)
    }

    /// Initialize with URL type for base URL.
    public static func initialize(
        apiKey: String,
        baseURL: URL,
        environment: SDKEnvironment = .production
    ) throws {
        let params = try SDKInitParams(apiKey: apiKey, baseURL: baseURL, environment: environment)
        try performCoreInit(with: params, startBackgroundServices: true)
    }

    // MARK: - Phase 1: Core Initialization (delegated to C++)

    private static func performCoreInit(with params: SDKInitParams, startBackgroundServices: Bool) throws {
        try coreInitQueue.sync {
            try performCoreInitSerial(with: params, startBackgroundServices: startBackgroundServices)
        }
    }

    private static func performCoreInitSerial(with params: SDKInitParams, startBackgroundServices: Bool) throws {
        guard !isInitializedFlag else { return }

        let initStartTime = CFAbsoluteTimeGetCurrent()

        // Set environment first so logging boots with correct config.
        state.withLock {
            $0.currentEnvironment = params.environment
            $0.initParams = params
            $0.hasCompletedServicesInit = false
            $0.hasCompletedHTTPSetup = false
            $0.httpSetupApplicable = true
            $0.servicesInitTask = nil
        }
        Logging.shared.applyEnvironmentConfiguration(params.environment)

        // Bring up the core C++ bridges (platform adapter, events,
        // telemetry, device callbacks). Must run before Phase 1 proto so
        // every C++ log routes through SDKLogger and analytics callbacks
        // are wired up.
        CppBridge.initialize(environment: params.environment)

        // Lifecycle INITIALIZATION_STAGE_* events (incl. duration_ms) are
        // published once by commons from rac_sdk_init_phase1_proto; Swift no
        // longer hand-emits duplicates.
        let logger = SDKLogger(category: "RunAnywhere.Init")

        do {
            // Configure C++ model-paths base directory before any
            // registerModel() calls so rac_model_registry_save() can
            // reconcile entries against on-disk folders inline.
            if let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {
                do {
                    try CppBridge.ModelPaths.setBaseDirectory(documentsURL)
                } catch {
                    logger.warning("Failed to set model paths base directory: \(error.localizedDescription)")
                }
            }

            // Phase 1 proto: validates inputs and runs rac_state_initialize.
            try CppBridge.SdkInit.phase1(
                environment: params.environment,
                apiKey: params.apiKey,
                baseURL: params.baseURL.absoluteString,
                deviceId: CppBridge.Device.persistentId
            )

            // SDK config (rac_sdk_init) + Keychain auth-storage install.
            // Idempotent state re-init is harmless; this call also wires up
            // version/platform metadata that Phase 1 proto does not touch.
            CppBridge.State.initialize(
                environment: params.environment,
                apiKey: params.apiKey,
                baseURL: params.baseURL,
                deviceId: CppBridge.Device.persistentId
            )

            state.withLock { $0.isInitialized = true }

            let initDurationMs = (CFAbsoluteTimeGetCurrent() - initStartTime) * 1000
            logger.info("Phase 1 complete in \(String(format: "%.1f", initDurationMs))ms (\(params.environment.description))")

            if isLocalOnlyMode {
                state.withLock {
                    $0.hasCompletedServicesInit = true
                    $0.hasCompletedHTTPSetup = false
                    $0.httpSetupApplicable = false
                }
                logger.debug("Phase 2 skipped for local-only Swift process")
            } else if startBackgroundServices {
                logger.debug("Starting Phase 2 (services) in background...")
                Task.detached(priority: .userInitiated) {
                    do {
                        try await completeServicesInitialization()
                        SDKLogger(category: "RunAnywhere.Init").info("Phase 2 complete (background)")
                    } catch {
                        SDKLogger(category: "RunAnywhere.Init")
                            .warning("Phase 2 failed (non-critical): \(error.localizedDescription)")
                    }
                }
            }

        } catch {
            logger.error("Initialization failed: \(error.localizedDescription)")
            state.withLock {
                $0.initParams = nil
                $0.currentEnvironment = nil
                $0.isInitialized = false
                $0.hasCompletedServicesInit = false
                $0.hasCompletedHTTPSetup = false
                $0.httpSetupApplicable = true
                $0.servicesInitTask = nil
            }
            throw error
        }
    }

    // MARK: - Phase 2: Services Initialization (Async)

    /// Complete services initialization (Phase 2). Safe to call multiple
    /// times; concurrent callers share the same Task so the step list runs
    /// at most once.
    public static func completeServicesInitialization() async throws {
        if hasCompletedServicesInit { return }

        let task: Task<Void, Error> = state.withLock {
            if let existingTask = $0.servicesInitTask {
                return existingTask
            }
            let newTask = Task<Void, Error> { try await _performServicesInitialization() }
            $0.servicesInitTask = newTask
            return newTask
        }

        do {
            try await task.value
            state.withLock { $0.servicesInitTask = nil }
        } catch {
            state.withLock { $0.servicesInitTask = nil }
            throw error
        }
    }

    /// Phase 2 step list. Commons owns the deterministic orchestration
    /// (auth through the registered HTTP transport, device registration,
    /// assignment fetch, telemetry flush, and downloaded-model discovery).
    /// Swift retains only platform-service callback registration.
    private static func _performServicesInitialization() async throws {
        let snapshot = state.withLock { ($0.initParams, $0.currentEnvironment) }
        guard let params = snapshot.0, let environment = snapshot.1 else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }

        let logger = SDKLogger(category: "RunAnywhere.Services")

        // Step 1: configure the Swift HTTP adapter used by callback-based
        // platform services. Auth and control-plane orchestration stay in C++.
        if await !CppBridge.HTTP.shared.isConfigured {
            if environment == .development {
                if await CppBridge.DevConfig.configureHTTP() {
                    logger.debug("HTTP adapter configured from C++ development config")
                } else {
                    logger.debug("HTTP adapter disabled: no usable development config")
                }
            } else if CppBridge.DevConfig.isUsableCredential(params.apiKey),
                      CppBridge.DevConfig.isUsableHTTPURL(params.baseURL.absoluteString) {
                await CppBridge.HTTP.shared.configure(baseURL: params.baseURL, apiKey: params.apiKey)
            } else {
                logger.debug("HTTP adapter disabled: no usable external config")
            }
        }

        // Step 2 (MainActor — must stay in Swift): platform-plugin and callback
        // registration. Commons uses these callbacks during Phase 2.
        await MainActor.run { CppBridge.initializeServices() }

        // Step 3 (C++): auth, device registration, model assignments,
        // telemetry flush, and downloaded-model discovery.
        let phase2Result = try CppBridge.SdkInit.phase2(
            buildToken: environment == .development ? CppBridge.DevConfig.buildToken : nil,
            forceRefreshAssignments: false,
            flushTelemetry: true,
            discoverDownloadedModels: true,
            rescanLocalModels: true
        )
        let completedHTTPSetup = phase2Result.hasCompletedHTTPSetup_p || phase2Result.httpConfigured
        if !phase2Result.warning.isEmpty {
            logger.info("Phase 2 warning: \(phase2Result.warning)")
        }
        if phase2Result.linkedModelsCount > 0 {
            logger.info("Phase 2 linked \(phase2Result.linkedModelsCount) assigned models")
        }

        state.withLock {
            $0.hasCompletedHTTPSetup = completedHTTPSetup
            $0.httpSetupApplicable = phase2Result.httpApplicable
            $0.hasCompletedServicesInit = true
        }
    }

    /// Ensure services are ready before API calls (internal guard).
    /// O(1) after first successful initialization with HTTP configured.
    /// If core services are done but HTTP/auth failed (offline init), retries auth only.
    internal static func ensureServicesReady() async throws {
        if isLocalOnlyMode && isInitializedFlag {
            return
        }

        let readiness = state.withLock {
            (
                services: $0.hasCompletedServicesInit,
                http: $0.hasCompletedHTTPSetup,
                applicable: $0.httpSetupApplicable
            )
        }
        if readiness.services && (readiness.http || !readiness.applicable) {
            return
        }
        if readiness.services && !readiness.http && readiness.applicable {
            await retryHTTPSetup()
            return
        }
        try await completeServicesInitialization()
    }

    /// Retry HTTP/auth after an offline initialization. Commons performs the
    /// round-trip through the registered platform HTTP transport.
    private static func retryHTTPSetup() async {
        guard currentEnvironment != nil else { return }
        let logger = SDKLogger(category: "RunAnywhere.HTTPRetry")

        let proto: RASdkInitResult
        do {
            proto = try CppBridge.SdkInit.retryHTTP()
        } catch {
            logger.debug("HTTP retry proto failed: \(error.localizedDescription)")
            return
        }

        let completedHTTPSetup = proto.hasCompletedHTTPSetup_p || proto.httpConfigured
        state.withLock {
            $0.hasCompletedHTTPSetup = completedHTTPSetup
            $0.httpSetupApplicable = proto.httpApplicable
        }

        if !proto.warning.isEmpty {
            logger.debug("HTTP retry warning: \(proto.warning)")
        }

        if completedHTTPSetup {
            logger.info("HTTP/Auth setup succeeded on retry")
        }
    }
}

File: sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/public/RunAnywhere.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * The main entry point for the RunAnywhere SDK.
 * Two-phase initialization is owned by commons (rac_sdk_init.h, P2-T9):
 *   * Phase 1 → rac_sdk_init_phase1_proto (validate + state init)
 *   * Phase 2 → rac_sdk_init_phase2_proto (device registration, model
 *     assignments, HTTP-state snapshot)
 *   * HTTP retry → rac_sdk_retry_http_proto
 * Kotlin retains only the parts that cannot move into C++:
 *   * Coroutine Mutex + servicesMutex concurrency primitive
 *   * EncryptedSharedPreferences / file-backed SDK params persistence
 *   * JNI platform-plugin registration on JVM/Android main thread
 *   * HTTP authentication round-trip via OkHttp (deferred per
 *     sdk_init.cpp file header)
 *   * Telemetry flush + model discovery (deferred — handles owned by SDK)
 */

package com.runanywhere.sdk.public

import com.runanywhere.sdk.infrastructure.logging.SDKLogger
import com.runanywhere.sdk.public.configuration.SDKEnvironment
import com.runanywhere.sdk.public.configuration.wireString
import com.runanywhere.sdk.public.events.EventBus
import com.runanywhere.sdk.public.extensions.LogLevel
import com.runanywhere.sdk.foundation.constants.SDKConstants
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

// ═══════════════════════════════════════════════════════════════════════════
// SDK INITIALIZATION FLOW (Two-Phase Pattern)
// ═══════════════════════════════════════════════════════════════════════════
//
// PHASE 1: Core Init (Synchronous, ~1-5ms, No Network)
// ─────────────────────────────────────────────────────
//   RunAnywhere.initialize(environment)
//     ├─ CppBridge.initialize()
//     │    ├─ PlatformAdapter.register()  ← File ops, logging, keychain
//     │    ├─ Events.register()           ← Analytics callback
//     │    ├─ Telemetry.initialize()      ← HTTP callback
//     │    └─ Device.register()           ← Device registration
//     └─ Mark: isInitialized = true
//
// PHASE 2: Services Init (Async, ~100-500ms, Network May Be Required)
// ────────────────────────────────────────────────────────────────────
//   RunAnywhere.completeServicesInitialization()
//     ├─ CppBridge.initializeServices()
//     │    ├─ ModelAssignment.register()  ← Model assignment callbacks
//     │    └─ Platform.register()         ← LLM/TTS service callbacks
//     └─ Mark: areServicesReady = true
//
// ═══════════════════════════════════════════════════════════════════════════

/**
 * The RunAnywhere SDK - Single entry point for on-device AI
 *
 * This object mirrors the iOS RunAnywhere enum pattern, providing:
 * - SDK initialization (two-phase: fast sync + async services)
 * - State access (isInitialized, areServicesReady, version, environment)
 * - Event access via `events` property
 *
 * Feature-specific APIs are available through extension functions in public/extensions/:
 * - STT: RunAnywhere.transcribe(), RunAnywhere.transcribeStream()
 * - TTS: RunAnywhere.synthesize(), RunAnywhere.loadModel(RAModelLoadRequest)
 * - LLM: RunAnywhere.generate(), RunAnywhere.generateStream()
 * - VAD: RunAnywhere.detectSpeech()
 * - VoiceAgent: VoiceAgentStreamAdapter(handle).stream() (v3.1)
 *
 * All AI component logic (LLM, STT, TTS, VAD) is delegated to the C++ runanywhere-commons
 * layer via CppBridge. Kotlin only handles platform-specific operations (HTTP, audio, file I/O).
 */
object RunAnywhere {
    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Private State
    // ═══════════════════════════════════════════════════════════════════════════

    private val logger = SDKLogger("RunAnywhere")

    @Volatile
    private var _currentEnvironment: SDKEnvironment? = null

    @Volatile
    private var _isInitialized: Boolean = false

    @Volatile
    private var _areServicesReady: Boolean = false

    private val lock = Any()
    private val servicesMutex = Mutex()

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Public Properties
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Check if SDK is initialized (Phase 1 complete)
     */
    val isInitialized: Boolean
        get() = _isInitialized

    /**
     * Check if services are fully ready (Phase 2 complete)
     */
    val areServicesReady: Boolean
        get() = _areServicesReady

    /**
     * Check if SDK is active and ready for use
     */
    val isActive: Boolean
        get() = _isInitialized && _currentEnvironment != null

    /**
     * Current SDK version
     */
    val version: String
        get() = SDKConstants.SDK_VERSION

    /**
     * Current environment (null if not initialized)
     */
    val environment: SDKEnvironment?
        get() = _currentEnvironment

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Event Access
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Event bus for SDK event subscriptions.
     *
     * Example usage:
     * ```kotlin
     * RunAnywhere.events.llmEvents.collect { event ->
     *     println("LLM event: ${event.type}")
     * }
     * ```
     */
    val events: EventBus
        get() = EventBus

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Authentication Info (Production/Staging only)
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Get the current user ID from authentication state.
     *
     * @return User ID if authenticated, `null` otherwise.
     */
    fun getUserId(): String? = platformGetUserId()

    /**
     * Get the current organization ID from authentication state.
     *
     * @return Organization ID if authenticated, `null` otherwise.
     */
    fun getOrganizationId(): String? = platformGetOrganizationId()

    /**
     * Check if the SDK is currently authenticated with a valid token.
     *
     * Equivalent to Swift's `RunAnywhere.isAuthenticated` static var.
     */
    val isAuthenticated: Boolean
        get() = platformIsAuthenticated()

    /**
     * Check if this device is registered with the backend.
     *
     * @return true if the device-registration handshake completed successfully.
     */
    fun isDeviceRegistered(): Boolean = platformIsDeviceRegistered()

    /**
     * The persistent device ID. Survives reinstalls (stored in keychain on
     * Apple platforms / EncryptedSharedPreferences on Android).
     *
     * Resolved by commons via the device-identity chain
     * (secure_get → vendor ID → freshly synthesized UUID).
     */
    val deviceId: String
        get() = platformDeviceId()

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Phase 1: Core Initialization (Synchronous)
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Initialize the RunAnywhere SDK (Phase 1)
     *
     * This performs fast synchronous initialization. Services initialization
     * is done separately via [completeServicesInitialization].
     *
     * ## Usage Examples
     *
     * ```kotlin
     * // Development mode (default)
     * RunAnywhere.initialize()
     *
     * // Production mode
     * RunAnywhere.initialize(environment = SDKEnvironment.SDK_ENVIRONMENT_PRODUCTION)
     * ```
     *
     * @param apiKey API key (optional for development, required for production/staging)
     * @param baseURL Backend API base URL (optional)
     * @param environment SDK environment (default: DEVELOPMENT)
     */
    fun initialize(
        apiKey: String? = null,
        baseURL: String? = null,
        environment: SDKEnvironment = SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT,
    ) {
        synchronized(lock) {
            if (_isInitialized) {
                logger.info("SDK already initialized")
                return
            }

            val initStartTime = System.currentTimeMillis()

            try {
                // Store environment
                _currentEnvironment = environment

                // Set log level based on environment
                val logLevel =
                    when (environment) {
                        SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT -> LogLevel.DEBUG
                        SDKEnvironment.SDK_ENVIRONMENT_STAGING -> LogLevel.INFO
                        SDKEnvironment.SDK_ENVIRONMENT_PRODUCTION -> LogLevel.WARNING
                        SDKEnvironment.SDK_ENVIRONMENT_UNSPECIFIED -> LogLevel.DEBUG
                    }
                SDKLogger.setLevel(logLevel)

                // Initialize CppBridge (Phase 1)
                // Note: CppBridge is in jvmAndroidMain, we call it via expect/actual
                initializeCppBridge(environment, apiKey, baseURL)

                // Mark Phase 1 complete
                _isInitialized = true

                val initDurationMs = System.currentTimeMillis() - initStartTime
                logger.info("✅ Phase 1 complete in ${initDurationMs}ms (${environment.wireString})")
            } catch (error: Exception) {
                logger.error("❌ Initialization failed: ${error.message}")
                _currentEnvironment = null
                _isInitialized = false
                throw error
            }
        }
    }

    /**
     * Initialize SDK for development mode (convenience method)
     */
    fun initializeForDevelopment(apiKey: String? = null) {
        initialize(apiKey = apiKey, environment = SDKEnvironment.SDK_ENVIRONMENT_DEVELOPMENT)
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - Phase 2: Services Initialization (Async)
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Complete services initialization (Phase 2)
     *
     * Called automatically on first API call, or can be awaited directly.
     * Safe to call multiple times - returns immediately if already done.
     */
    suspend fun completeServicesInitialization() {
        // Fast path: already completed
        if (_areServicesReady) {
            return
        }

        servicesMutex.withLock {
            if (_areServicesReady) {
                return
            }

            if (!_isInitialized) {
                throw IllegalStateException("SDK must be initialized before completing services initialization")
            }

            logger.info("Initializing services for ${_currentEnvironment?.wireString} mode...")

            try {
                // Initialize CppBridge services (Phase 2)
                initializeCppBridgeServices()

                // Mark Phase 2 complete
                _areServicesReady = true

                logger.info("✅ Services initialized for ${_currentEnvironment?.wireString} mode")
            } catch (e: Exception) {
                logger.error("Services initialization failed: ${e.message}")
                throw e
            }
        }
    }

    /**
     * Ensure services are ready before API calls (internal guard).
     *
     * Mirrors Swift `RunAnywhere.ensureServicesReady()`. Lazily kicks off
     * Phase 2 (`completeServicesInitialization()`) the first time a feature
     * entry is hit, so commonMain consumers do **not** need to make that
     * call explicitly — the same shape as the iOS SDK's `_servicesInitTask`
     * fan-in. The Mutex guard inside [completeServicesInitialization]
     * serializes concurrent first-callers.
     *
     * O(1) after the first successful initialization.
     *
     * @throws IllegalStateException if Phase 1 ([initialize]) has not run.
     */
    internal suspend fun ensureServicesReady() {
        if (_areServicesReady) {
            return // O(1) fast path — Phase 2 already complete
        }
        // Phase 1 must be complete before we can lazily start Phase 2.
        requireInitialized()
        // Auto-start Phase 2 — `completeServicesInitialization` is mutex-guarded
        // and short-circuits if another caller already finished the work.
        completeServicesInitialization()
    }

    /**
     * Ensure SDK is initialized (throws if not)
     */
    internal fun requireInitialized() {
        if (!_isInitialized) {
            throw IllegalStateException("SDK not initialized. Call RunAnywhere.initialize() first.")
        }
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - SDK Reset
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Reset SDK state
     * Clears all initialization state and releases resources
     */
    suspend fun reset() {
        logger.info("Resetting SDK state...")

        synchronized(lock) {
            // Shutdown CppBridge
            shutdownCppBridge()

            _isInitialized = false
            _areServicesReady = false
            _currentEnvironment = null
        }

        logger.info("SDK state reset completed")
    }

    /**
     * Cleanup SDK resources without full reset
     */
    suspend fun cleanup() {
        logger.info("Cleaning up SDK resources...")
        // Cleanup logic here
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // MARK: - CppBridge Integration (expect/actual pattern)
    // ═══════════════════════════════════════════════════════════════════════════

    /**
     * Initialize CppBridge (Phase 1)
     * Implementation is in jvmAndroidMain via expect/actual
     */
    private fun initializeCppBridge(environment: SDKEnvironment, apiKey: String?, baseURL: String?) {
        logger.debug("CppBridge initialization requested for $environment")
        initializePlatformBridge(environment, apiKey, baseURL)
    }

    /**
     * Initialize CppBridge services (Phase 2)
     * Implementation is in jvmAndroidMain via expect/actual
     */
    private suspend fun initializeCppBridgeServices() {
        logger.debug("CppBridge services initialization requested")
        initializePlatformBridgeServices()
    }

    /**
     * Shutdown CppBridge
     * Implementation is in jvmAndroidMain via expect/actual
     */
    private fun shutdownCppBridge() {
        logger.debug("CppBridge shutdown requested")
        shutdownPlatformBridge()
    }
}

// ═══════════════════════════════════════════════════════════════════════════
// Platform-specific bridge functions (expect/actual pattern)
// ═══════════════════════════════════════════════════════════════════════════

/**
 * Initialize platform-specific bridge (Phase 1).
 * On JVM/Android, this calls CppBridge.initialize() to load native libraries.
 *
 * @param environment SDK environment
 * @param apiKey API key for authentication (required for production/staging)
 * @param baseURL Backend API base URL (required for production/staging)
 */
internal expect fun initializePlatformBridge(environment: SDKEnvironment, apiKey: String?, baseURL: String?)

/**
 * Initialize platform-specific bridge services (Phase 2).
 * On JVM/Android, this calls CppBridge.initializeServices().
 */
internal expect suspend fun initializePlatformBridgeServices()

/**
 * Shutdown platform-specific bridge.
 * On JVM/Android, this calls CppBridge.shutdown().
 */
internal expect fun shutdownPlatformBridge()

// ═══════════════════════════════════════════════════════════════════════════
// Platform-specific auth/device state accessors (expect/actual pattern)
// ═══════════════════════════════════════════════════════════════════════════

/** JVM/Android actual routes to `RunAnywhereBridge.racAuthGetUserId()`. */
internal expect fun platformGetUserId(): String?

/** JVM/Android actual routes to `RunAnywhereBridge.racAuthGetOrganizationId()`. */
internal expect fun platformGetOrganizationId(): String?

/** JVM/Android actual routes to `RunAnywhereBridge.racAuthIsAuthenticated()`. */
internal expect fun platformIsAuthenticated(): Boolean

/** JVM/Android actual routes to `CppBridgeDevice.isRegistered()`. */
internal expect fun platformIsDeviceRegistered(): Boolean

/**
 * JVM/Android actual routes to `CppBridgeDevice.getDeviceId()` →
 * `RunAnywhereBridge.racAuthGetDeviceId()` chain, returning empty string when
 * the device hasn't been registered yet.
 */
internal expect fun platformDeviceId(): String
